Improve scarecrow

ShadowProtocol opened the pull request:
Implements #1855 

# Changes proposed in this pull request:
- Scarecrow (Field) now physically takes up two blocks
- Scarecrow GUI is accessible from either half of scarecrow 
- Fixed a glitch where dragging a seed from the field slot would duplicate the seed
    due to the wrong inventory handler
- Modified the scarecrow held item to conform slightly closer to the art style of similar held items

Please review.

ShadowProtocol committed:
Add upper half of scarecrow

Add a double block blockstate to introduce a secondary UPPER scarecrow block
Update place, destroy, and activate logic in scarecrow block to consider halves
Update the tile entity placement and field container to only use the LOWER half

ShadowProtocol committed:
Fix scarecrow inventory

Replace the IItemHandler with the less glitchy IInventory
Simplify shift-click transfer function
Add Slot anonymous extension for validating seeds

ShadowProtocol committed:
Adjust scarecrow hitbox and initial rotation

Modify placement rotation to face the player
Make both halves of scarecrow block reflect the same hitbox

ShadowProtocol committed:
Change scarecrow held item model

Set scarecrow held item to be generated as an item with rotation, not a built-in

ShadowProtocol committed:
Adjust scarecrow held item texture

ShadowProtocol committed:
Fix scarecrow placement condition

Ensure the upper block of the scarecrow has a clear space to be placed in

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ldtteam/minecolonies?pullRequest=5241) <br/>All committers have signed the CLA.

Raycoms commented on src/main/java/com/minecolonies/coremod/tileentities/ScarecrowTileEntity.java:
```diff
@@ -119,87 +117,16 @@
     /**
      * Inventory of the field.
      */
-    private final IItemHandlerModifiable inventory = new IItemHandlerModifiable() {
```
As far as I know we should not be using IInventory anymore and instead use IItemhandlers.

ShadowProtocol replied:
Ah, true true. There was still one of the functions causing a glitch in the implementation that was there before. I believe it is fixed now. Not quite sure which part was causing it still.

Raycoms commented on src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java:
```diff
@@ -135,17 +167,27 @@ public void onExplosionDestroy(final World worldIn, final BlockPos pos, final Ex
+
+        // just double check the other block is also the scarecrow and not the same half,
+        // then destroy it (make it air)
+        if (otherstate.getBlock() == this && otherstate.get(HALF) != half) {
```
formatting
{

}

(All over the PR)

Raycoms commented on src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java:
```diff
@@ -101,8 +114,23 @@ public ActionResultType onBlockActivated(
+        }
+    }
+
+    public boolean isValidPosition(BlockState state, IWorldReader worldIn, BlockPos pos) {
```
@Override

ShadowProtocol committed:
Use ItemStackHandler in scarecrow tile entity

Switch IInventory for Forge IItemHandler's default implementation ItemStackHandler
Fix brace formatting
Add Override annotations

ShadowProtocol committed:
Fix placement glitch with scarecrow

Fix other interactions like block placement from starting when the scarecrow is interacted with

Asherslab commented on src/api/java/com/minecolonies/api/inventory/container/ContainerField.java:
```diff
@@ -87,28 +97,28 @@ public ContainerField(final int windowId, final PlayerInventory playerInventory,
+            if (!mergeItemStack(transfer, 1, 37, true)) {
+                return ItemStack.EMPTY;
+            }
+        } else {
```
formatting

Asherslab commented on src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java:
```diff
@@ -87,22 +95,49 @@ public ActionResultType onBlockActivated(
             {
                 NetworkHooks.openGui((ServerPlayerEntity) player, (ScarecrowTileEntity) entity, pos);
-                return ActionResultType.SUCCESS;
+            } else {
```
formatting

Asherslab commented on src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java:
```diff
@@ -87,22 +95,49 @@ public ActionResultType onBlockActivated(
+        if (context.getPos().getY() < 255 && context.getWorld().getBlockState(context.getPos().up()).isReplaceable(context))
+        {
+            return this.getDefaultState().with(FACING, Direction).with(HALF, DoubleBlockHalf.LOWER);
+        } else {
```
formatting

Asherslab commented on src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java:
```diff
@@ -87,22 +95,49 @@ public ActionResultType onBlockActivated(
+        if (state.get(HALF) == DoubleBlockHalf.LOWER)
+        {
+            return blockstate.isSolidSide(worldIn, blockpos, Direction.UP);
+        } else {
```
formatting

Asherslab commented on src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java:
```diff
@@ -135,17 +176,28 @@ public void onExplosionDestroy(final World worldIn, final BlockPos pos, final Ex
     public void onBlockHarvested(final World worldIn, @NotNull final BlockPos pos, final BlockState state, @NotNull final PlayerEntity player)
     {
+        DoubleBlockHalf half = state.get(HALF);
+        BlockPos otherpos = half == DoubleBlockHalf.LOWER? pos.up() : pos.down();
```
formatting (space between LOWER and `?` in ternaries)

Asherslab commented on src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java:
```diff
@@ -135,17 +176,28 @@ public void onExplosionDestroy(final World worldIn, final BlockPos pos, final Ex
     }
-
+/*
     @Override
```
please remove code don't comment it out

Asherslab commented on src/main/java/com/minecolonies/coremod/tileentities/ScarecrowTileEntity.java:
```diff
@@ -578,7 +518,7 @@ public void read(final CompoundNBT compound)
             if (ItemStackUtils.getSize(stack) <= 0)
             {
-                inventory.setStackInSlot(i, ItemStackUtils.EMPTY);
+                inventory.setStackInSlot(i, ItemStack.EMPTY);
```
usage of `ItemStackUtils` version is preferred for easier future porting (reason it was made in the first place)

Asherslab reviewed (dismissed):
looks good other than some cleanup

ShadowProtocol committed:
Fix formatting

Restore uses of ItemStackUtils.Empty
Expand else statements and ternary statements

Raycoms committed:
Merge branch 'version/1.15' into feature/scarecrow

Raycoms commented:
:shipit: 

[![Approved with PullApprove](https://img.shields.io/badge/maintainers-approved-brightgreen.svg)](https://pullapprove.com/ldtteam/minecolonies/pull-request/5241/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ldtteam/minecolonies)

Strohtaler commented:
Great Job, working fine! and JIT, I was about to create an issue about the dragging seeds problem [: -)
